fog/ledger/server/src/key_image_service.rs

```diff
@@ -147,30 +149,42 @@ impl<L: Ledger + Clone, E: LedgerEnclaveProxy> KeyImageService<L, E> {
         queries: Vec<attest::NonceMessage>,
     ) -> MultiKeyImageStoreResponse {
         let mut response = MultiKeyImageStoreResponse::new();
+        let uri = fog_ledger_store_uri;
         // The router needs our own URI, in case auth fails / hasn't been started yet.
-        response.set_store_uri(fog_ledger_store_uri.url().to_string());
+        response.set_store_uri(uri.url().to_string());
         // Default status of AUTHENTICATION_ERROR in case of empty queries
         response.set_status(MultiKeyImageStoreResponseStatus::AUTHENTICATION_ERROR);
 
+        let start_time = Instant::now();
+        let subdomain = uri.subdomain().unwrap_or("");
+        let histogram_observe = |status: &str| {
+            let histogram = QUERY_REQUESTS.with_label_values(&[subdomain, status]);
+            histogram.observe(start_time.elapsed().as_secs_f64());
+        };
+
         for query in queries.into_iter() {
             // Only one of the query messages in the multi-store query is intended for this
             // store. It's a bit of a broadcast model - all queries are sent to
             // all stores, and then the stores evaluate which message is meant
             // for them.
             match self.check_key_image_store_auth(query) {
                 Ok(attested_message) => {
+                    histogram_observe("success");
```

Comment:
It seems odd to me that the start time is captured before the loop, but we are reporting per-query timing in each iteration. That means that each observation will also count the time it took for all queries that happened before it.